neuralmonkey/decoders/transformer.py

```diff
@@ -143,101 +143,125 @@ def embedded_train_inputs(self) -> tf.Tensor:
                 self_context, self.dropout_keep_prob, self.train_mode)
 
-    def encoder_attention(self, level: int, queries: tf.Tensor) -> tf.Tensor:
+            # Residual connections
+            return self_context + prev_layer.temporal_states
+
+    def encoder_attention_sublayer(
+            self, queries: tf.Tensor) -> tf.Tensor:
 
-        with tf.variable_scope("dec_inter_att_level_{}".format(level),
+        with tf.variable_scope("encdec_attention",
                                reuse=tf.AUTO_REUSE):
             encoder_att_states = get_attention_states(self.encoder)
             encoder_att_mask = get_attention_mask(self.encoder)
 
+            # Layer normalization
+            normalized_queries = tf.contrib.layers.layer_norm(
+                queries, begin_norm_axis=2)
+
             # TODO handle histories
+            # Attend to the encoder
             encoder_context, _ = attention(
-                queries=queries,
+                queries=normalized_queries,
                 keys=encoder_att_states,
                 values=encoder_att_states,
                 keys_mask=encoder_att_mask,
                 num_heads=self.n_heads_enc,
                 dropout_callback=lambda x: dropout(
                     x, self.attention_dropout_keep_prob, self.train_mode))
 
-            return dropout(
+            encoder_context = dropout(
                 encoder_context, self.dropout_keep_prob, self.train_mode)
 
+            # Residual connections
+            return encoder_context + queries
+
+    def feedforward_sublayer(
+            self, layer_input: tf.Tensor) -> tf.Tensor:
+
+        with tf.variable_scope("feedforward"):
+            # Layer normalization
+            normalized_input = tf.contrib.layers.layer_norm(
+                layer_input, begin_norm_axis=2)
+
+            # Feed-forward network hidden layer + ReLU + dropout
+            ff_hidden = tf.layers.dense(normalized_input,
+                                        self.ff_hidden_size,
+                                        activation=tf.nn.relu,
+                                        name="hidden_state")
+            ff_hidden = dropout(ff_hidden,
+                                self.dropout_keep_prob,
+                                self.train_mode)
+
+            # Feed-forward output projection + dropout
+            ff_output = tf.layers.dense(ff_hidden,
+                                        self.dimension,
+                                        name="output")
+            ff_output = dropout(ff_output,
+                                self.dropout_keep_prob,
+                                self.train_mode)
+
+            # Residual connections
+            return ff_output + layer_input
+
     def layer(self, level: int, inputs: tf.Tensor,
               mask: tf.Tensor) -> TransformerLayer:
-        # Recursive implementation. Outputs of the zeroth layer are the inputs
+        # Recursive implementation. Outputs of the zeroth layer
```

Comment:
Možná to sem nepatří, ale zřídil bych nějaké `self.layers`, kam by layery appendovaly, abychom je měli někde dobře přístupně uložené, až budeme chtít nějak slušně vizualizovat ten self-attention.

Reply:
Udelal bych na to issue.

Reply:
tak ju udělej